services/calculateBlockTvl.js

```diff
@@ -0,0 +1,149 @@
+            }
+        }
+    } else {
+        log(chalk.blue.bold('ALREADY RUN EXTRACTION FOR THIS BLOCK, returning ') + `#${header}`);
+    }
+};
+
+async function extractBlockNumber(blockNumber) {
+    const blockHash = await karuraApi.rpc.chain.getBlockHash(blockNumber);
+    const signedBlock = await karuraApi.rpc.chain.getBlock(blockHash);
+    await extractBlock(signedBlock.block, blockNumber);
+}
+
+async function extractBlock(block, blockNumber) {
+    log(chalk.blue.bold('API: ') + `Extracting Block: ${block.header.hash}`);
+
+    const timestamp = await karuraApi.query.timestamp.now.at(block.header.hash);
+    log(chalk.blue.bold('API: ') + `Timestamp: ${timestamp.toString()}`);
+
+    await extractBlockDexLiquidities(block, timestamp.toString(), blockNumber);
+    // await extractBlockEvents(block, timestamp);
+}
+
+async function extractBlockDexLiquidities(block, timestamp, blockNumber) {
+    const tradingPairs = config.TOKEN_PAIRS.map((pair) => {
+        const tokenA = karuraApi.registry.createType('CurrencyId', { Token: pair[0] });
+        const tokenB = karuraApi.registry.createType('CurrencyId', { Token: pair[1] });
+        const tradingPair = new TokenPair(Token.fromCurrencyId(tokenA), Token.fromCurrencyId(tokenB)).toTradingPair(karuraApi) 
+        return new Promise((resolve, reject) => {
+            karuraApi.query.dex.liquidityPool.at(block.header.hash, tradingPair)
+            .then(res => {
+                resolve({
+                    pair: pair,
+                    data: res
+                })
+            })
+            .catch(e => {
+                reject(e)
+            });
+        });
+    });
+    await extractBlockValue(tradingPairs, timestamp, blockNumber);
+}
+
+async function extractBlockValue(tradingPairs, timestamp, blockNumber) {
+    const dexBalances = await Promise.all(tradingPairs);
+    const liquidity = dexBalances.map((balance) => {
+        return {
+            data: {
+                method: "dex.Liquidity",
+                [balance.pair[0]]: balance.data[0].toString(),
+                [balance.pair[1]]: balance.data[1].toString(),
+                timestamp: timestamp
+            },
+            pair: balance.pair,
+        }
+    });
+
+    // Total Loan Positions
+    const loanBalances = await karuraApi.query.loans.totalPositions({ TOKEN: 'KSM' });
+    const loanPositions = {
+        timestamp: timestamp,
+        collateral: loanBalances.collateral.toString(),
+        debit: loanBalances.debit.toString(),
+    }
+
+    // Total Crowd Loan Funds
+    const crowdloan = await kusamaApi.query.crowdloan.funds(config.KUSAMA_CROWDLOAN_ID);
+    const fundsRaised = crowdloan.toString().raised;
+    
+    // Final Block Data
+    let blockTVL = new BlockTVL(liquidity, loanPositions, fundsRaised, blockNumber.toString())
+
+    // write lastBlock informaiton to DB
+    await recordBlockData(blockTVL);
+}
+
+async function recordBlockData (blockTVL) {   
+    await client.connect();
```

Comment:
is it necessary to connect every time?

Reply:
In this case we do. I suspect Mongo is only an interim solution here and migrating to InfluxDB would change the implementation anyway